okhttp/src/jvmMain/kotlin/okhttp3/internal/platform/android/AndroidDns.kt

```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright (C) 2022 Square, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package okhttp3.internal.platform.android
+
+import android.annotation.SuppressLint
+import android.net.DnsResolver
+import java.net.InetAddress
+import java.util.concurrent.Executors
+import java.util.concurrent.atomic.AtomicInteger
+import okhttp3.internal.dns.AsyncDns
+
+@SuppressLint("NewApi") // API 29!
+class AndroidDns : AsyncDns {
+  val resolver = DnsResolver.getInstance()
+  val executor = Executors.newSingleThreadExecutor()
+
+  override fun query(hostname: String, callback: AsyncDns.Callback) {
+    val executing = AtomicInteger(2)
+
+    resolver.query(
```

Comment:
Split by type, otherwise the results are nicely interleaved.

> Send a DNS query with the specified name and query type, get back a set of InetAddresses with rfc6724 sorting style asynchronously. The answer will be provided asynchronously through the provided [Callback](https://developer.android.com/reference/android/net/DnsResolver.Callback).

Reply:
This is the first part of Happy Eyeballs ordering - which is helpfully useful.

> First, the client MUST sort the addresses received up to this point using Destination Address Selection ([[RFC6724], Section 6](https://datatracker.ietf.org/doc/html/rfc6724#section-6)).